### PR TITLE
Harden circular dependency parser in lint:cycles (oh-tab-zhwi.1.3)

### PR DESCRIPTION
**Pull request by smolpaws**
## Summary
- harden `parseCycleJson` in `scripts/check-circular-deps.mjs` to use strict JSON parsing first
- add a bounded fallback parser that scans bracketed arrays safely (depth + string/escape handling)
- validate payload shape (`string[][]`) and surface actionable parse diagnostics when output is invalid
- add targeted unit tests for strict, empty, noisy embedded, multi-array, and invalid outputs

## Testing
- `npx vitest run src/__tests__/checkCircularDeps.test.ts`
- `npm run lint`
- `npm run typecheck`

### Bead
[
  {
    "id": "oh-tab-zhwi.1.3",
    "title": "Harden circular dependency parser in lint:cycles",
    "description": "Improve scripts/check-circular-deps.mjs parseCycleJson handling so it first parses strict JSON output and only falls back to bounded extraction with clear diagnostics. Current first-/last- fallback can be brittle when upstream output includes bracketed warnings.",
    "acceptance_criteria": "- parseCycleJson prefers strict JSON parsing and reports actionable errors.\\n- Fallback parser is bounded/sentinel-aware and covered by tests.\\n- lint:cycles behavior remains backward compatible for normal madge output.",
    "notes": "Started implementation: strict-first cycle JSON parsing, bounded fallback scanner, and parser unit tests.",
    "status": "in_progress",
    "priority": 2,
    "issue_type": "task",
    "created_at": "2026-02-05T23:04:22.908941+01:00",
    "updated_at": "2026-02-05T23:14:51.180853+01:00",
    "labels": [
      "architecture",
      "guardrails",
      "lint",
      "refactor"
    ],
    "dependencies": [
      {
        "id": "oh-tab-zhwi",
        "title": "Repo-wide architecture hardening and AI-debt cleanup",
        "description": "Program to audit and harden architecture across the full repository. Scope: extension host, webview UI/host bridge, shared/auth/settings/hal modules, and packages/agent-sdk-ts (runtime, conversation/context, llm, tools/workspace). Outcome: reduced complexity, explicit architectural boundaries, cycle/duplication reduction, and CI-enforced guardrails.",
        "acceptance_criteria": "- Every major component group has an analysis bead with findings and follow-up fix beads.\\n- Fix beads are prioritized and executed via branch+PR+review workflow.\\n- Architecture guardrails are enforced in lint/CI.\\n- No merges occur without explicit reviewer approval in Agent Mail thread.",
        "notes": "Baseline findings recorded and child beads created for cycles, duplication, App/eventBlocks decomposition, host-message registry refactor, LocalWorkspace/TerminalTool modularization, and RemoteConversation alignment (oh-tab-4pc). First PR opened for oh-tab-zhwi.1.1 (#947), pending reviewer approval.",
        "status": "in_progress",
        "priority": 0,
        "issue_type": "epic",
        "assignee": "RedCastle",
        "created_at": "2026-02-05T22:34:43.387869+01:00",
        "updated_at": "2026-02-05T22:42:53.977274+01:00",
        "labels": [
          "architecture",
          "audit",
          "tech-debt",
          "top-priority"
        ],
        "dependency_type": "discovered-from"
      },
      {
        "id": "oh-tab-zhwi.1",
        "title": "Architecture guardrails baseline (lint/cycle/duplication policies)",
        "description": "Define and implement architecture quality gates that prevent regressions. Seed from baseline findings: circular dependencies in src/webview/host and packages/agent-sdk-ts, large god-files, and production duplication clusters.",
        "acceptance_criteria": "- Proposal for lint/CI guardrails with staged rollout (warn-\u003eerror).\\n- At least one enforceable guardrail lands (or is captured as follow-up with clear rationale).\\n- Guardrails documented in repo docs/AGENTS references.",
        "status": "in_progress",
        "priority": 0,
        "issue_type": "task",
        "assignee": "RedCastle",
        "created_at": "2026-02-05T22:35:05.936425+01:00",
        "updated_at": "2026-02-05T22:35:37.683769+01:00",
        "labels": [
          "architecture",
          "audit",
          "ci",
          "lint"
        ],
        "dependency_type": "discovered-from"
      }
    ]
  }
]

---

<a href="https://app.devin.ai/review/enyst/openhands-tab/pull/950" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved circular dependency detection to handle various output formats with more robust parsing
  * Enhanced error messages for better diagnostics when issues occur

### CODE REVIEW

File: scripts/check-circular-deps.mjs [hunk 1]
```diff
@@ -34,19 +36,101 @@ const TARGETS = [
...
+    if (char === ']') {
+      depth -= 1;
+      if (depth === 0) {
+        return i;
+      }
+      if (depth < 0) {
+        return -1;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (smolpaws) [reply scripts/check-circular-deps.mjs:96]: Thanks for the readability suggestion. I’m keeping the current structure because each early `continue` maps directly to one parser state transition branch and keeps string/escape handling localized; behavior is equivalent, so I’m declining this as non-blocking for now.
